Single registered article guest view

brianteeman opened the pull request:
Pull Request for Issue #9830
#### Testing Instructions

Create a single article with an intro and full image and a readmore
Set the article to registered only
Create a menu link to the article of type single article
In the menu options make sure show unauthorised links is set to show

Now check the menu item on the front end - you should see the introtext but not the intro image and when you login you see all the text and the full image

Apply the patch
Now check the menu item on the front end - you should see the introtext and the intro image and when you login you see all the text and the full image

brianteeman committed:
Merge pull request #8 from joomla/staging

Update from upstream

brianteeman committed:
Update default.php

Webdongle commented:
I have tested this item :red_circle: unsuccessfully on 030de773fdb6dda9ac253d22998f6f23dc8700fb<br /><br />With  Show Voting set to 'Show' ... the voting dropdown box displays twice to guests
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

smz commented on components/com_content/views/article/tmpl/default.php:
```diff
@@ -133,7 +133,14 @@
 	<?php // Optional teaser intro text for guests ?>
 	<?php elseif ($params->get('show_noauth') == true && $user->get('guest')) : ?>
-	<?php echo $this->item->introtext; ?>
+    	<?php echo JLayoutHelper::render('joomla.content.intro_image', $this->item); ?>
```
Move down below the "if"

smz replied:
No, sorry, I was wrong on this: I mistook the if below as the if on "image_intro", but that's already in the JLayout

smz commented on components/com_content/views/article/tmpl/default.php:
```diff
@@ -133,7 +133,14 @@
+    	<?php echo JLayoutHelper::render('joomla.content.intro_image', $this->item); ?>
+	<?php if (!$params->get('show_intro')) : ?>
+	<?php // Content is generated by content plugin event "onContentAfterTitle" ?>
+	<?php echo $this->item->event->afterDisplayTitle; ?>
```
already done in line 85 ("voting" is here)
all which is above lie 95 should not be duplicated here

smz replied:
Also the handling of "afterDisplayTitle" does not belongs here as it has already been performed on lines 87-89 (_will generate duplicate_...)

smz commented:
proposed some changes
sorry for the brevity (on the road...)

brianteeman committed:
Update default.php

brianteeman commented:
Updated to remove double votes

rgmears commented:
Works for me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

woluweb commented:
I have tested this item :white_check_mark: successfully on dc144b3b84ceb21afbc23619b18901de82d486bc<br /><br />Tested successfully on [cc44ae1](https://github.com/joomla/joomla-cms/commit/cc44ae16209d260a0750069104bcac759bb37621 "cc44ae1")<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

wojsmol commented:
@rgmears Please mark test result on https://issues.joomla.org/tracker/joomla-cms/9865

rgmears commented:
@wojsmol 
>Please mark test result on ....

I don't know how to do that. Sorry.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

wojsmol commented:
@rgmears See https://docs.joomla.org/Testing_Joomla!_patches#Recording_test_results

brianteeman commented:
@rgmears @wojsmol its ok I manually added his test :)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

smz commented:
I cannot test now (_on a train with no test environ with me_) but by simple code review it is evident that that lines 137-140 are a duplicate of line 87.

Both echoes the content generated by "afterDisplayTitle" plugins (_probably a very seldom used event, but anyway..._) which in case will be duplicated...

In Joomla event lists (https://docs.joomla.org/Plugin/Events/Content) I don't find "afterDisplayTitle": is it the same as "onContentAfterTitle"?

smz commented:
In https://docs.joomla.org/Plugin/Events/Content a note in the "onContentAfterTitle" event description says that
> ... this event has special purpose in com_content for use in handling the introtext.

but I'm unsure of what this is or could be as no further specification is given.

What is evident from code is that content generated by this event is displayed only if the "show_intro" param is set.

smz commented:
Confirmed by testing:

#### Without:
![without](https://cloud.githubusercontent.com/assets/552475/14547347/f28ab770-02ac-11e6-8b91-0db086287770.png)

#### With:
![with](https://cloud.githubusercontent.com/assets/552475/14547350/fb24fb48-02ac-11e6-842d-ee0f1ee864ed.png)

The "Test Aftertitle" plugin is:
~~~
<?php
defined('_JEXEC') or die;

class PlgContentTest_aftertitle extends JPlugin {

	// onContentAfterTitle handler
	public function onContentAfterTitle($context, &$article, &$params, $limitstart) {
		return '<h3>This is generated by the "Test Aftertitle" plugin</h3>';
	}

}
~~~

smz commented:
I have tested this item :red_circle: unsuccessfully on dc144b3b84ceb21afbc23619b18901de82d486bc<br /><br />See above...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

smz committed:
Update default.php

This solves the duplication of the output for the "onContentAfterTitle" event and also solves the issues that the "onContentPrepare" event is not handled for the intro text.

I'm unsure, anyway, if it is correct to fix the latter here or if it would be better to fix it at the "view" level (_so that all templates, even overridden ones_) could benefit of the fix

brianteeman committed:
Merge pull request #12 from smz/patch-1

Update default.php

Webdongle commented:
As the patch has been updated because of the fail when tested with the edited plugin ... should the Testing Instructions be updated to take into account that part of the test ?


>In https://docs.joomla.org/Plugin/Events/Content a note in the "onContentAfterTitle" event description says that
>>    ... this event has special purpose in com_content for use in handling the introtext.

Has me totally confused and am unsure how to perform the test with the edited plugin ?

<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

smz commented:
@Webdongle 
TBH, I'm as much confused as you are, and probably quite more (_but I'll elaborate about that later on_)

As far as regards testing instructions the last word is of course @brianteeman's one, but I can explain what I did in the PR I proposed to Brian and he later incorporated, if this can help.

 1) I eliminated the duplicate display of output generated for the "onContentAfterTitle". This wasn't there at the beginning and was introduced (_by mistake, IMHO_) in Brian's PR. Eliminating this, thus, doesn't change the expected behaviour and hence testing instructions.

 2) I added an extra processing step so that the intro text is passed to the "onContentPrepare" event. This fixes the issue of directives like `{loadposition}` or `{loadmodule}` appearing verbatim in the generated output instead of being processed. I added this as I interpreted this was Brian wish when (_while closing #9830_) he said:

> ... better code than suggested before as it uses the layout correctly **and the plugins (i hope)** (_my emphasis_ )

**This second change requires of course also testing that those directives are indeed honoured** (_as they are, from my testing_).

About this second change I'm also unsure (_and I expressed my doubts directly to Brian_) if it is correct to do that here or would it better to do it here or it would be better to perform it at an higher level (_at the view level, probably_) so that not only this template, but also all other templates using the "intro text" would benefit of the fix.
___________
##### Unsure if off topic (_about the "onContentAfterTitle" event and the "Show Intro Text" option_).

The "Show intro text" option seems to be quite a strange beast. The description is:
> If set to Show, the Intro Text of the article will show when you drill down to the article. If set to Hide, only the part of the article after the "Read More" break will show.

It indeed does that (_intro text not displayed in the "Single Article" view if set to "Hide"_) **but it also controls (_with inverted logic_) weather the content generated by the "onContentAfterTitle" should be displayed or not.**

This is applied, apparently in an inconsistent way, throughout different views. For this, I think we'll need the memory of "_old timers_" to explain why this is done this way and what's the logic behind it (_I'm calling @infograf768 for that, as apparently he did put his hands on that and probably should know..._)

conconnl commented:
I have tested this item :white_check_mark: successfully on bc7d64c6c2f67d20db97de2483d2767d40d39a79<br /><br />Tested successful per instructions<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

Simonkloostra commented:
I have tested this item :white_check_mark: successfully on bc7d64c6c2f67d20db97de2483d2767d40d39a79<br /><br />Testscenario executed successfully, intro-image is now shown in single-article view<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

rgmears commented:
I have tested this item :white_check_mark: successfully. 

Intro Image shows. 

`{loadposition` and `{loadmodule}` both display content not code.

joris85 commented:
Tested successfully,
As quest you do not see the intro image while you should see the intro image.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

klatte88 commented:
Patch tested succesfully after patch intro image shows up as expected<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

klatte88 commented:
I have tested this item :white_check_mark: successfully on bc7d64c6c2f67d20db97de2483d2767d40d39a79<br /><br />problem as described recreated and solved after patch<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

brianteeman commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/9865">issues.joomla.org/joomla-cms/9865</a>.</sub>

rgmears commented:
Forgive the uninformed question, but, what does RTC mean?

smz commented:
@rgmears 
> Forgive the uninformed question, but, what does RTC mean?

**R**eady **T**o **C**ommit

@brianteeman 
So you think it is OK to fix the "onContentPrepare" issue here?

rgmears commented:
Thanks @smz .

brianteeman commented:
Create a new issue please

On 15 April 2016 at 15:04, Sergio Manzi <notifications@github.com> wrote:

> @rgmears <https://github.com/rgmears>
>
> Forgive the uninformed question, but, what does RTC mean?
>
> *R*eady *T*o *C*ommit
>
> @brianteeman <https://github.com/brianteeman>
> So you think it is OK to fix the "onContentPrepare" issue here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/9865#issuecomment-210475381>
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

smz commented:
about what?

smz commented:
... pinging all who were involved in this as I amended it in #9964: thanks for testing and/or review.
